File: src/network/client/ServerClientCache.php
<?php

namespace pocketcloud\cloud\network\client;

use Closure;
use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\network\Network;
use pocketcloud\cloud\network\packet\impl\KeepAlivePacket;
use pocketcloud\cloud\PocketCloud;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\server\CloudServerManager;
use pocketcloud\cloud\template\TemplateType;
use pocketcloud\cloud\util\misc\Tickable;
use pocketcloud\cloud\util\net\Address;
use pocketcloud\cloud\util\ProcessUtils;
use pocketcloud\cloud\util\trait\SingletonTrait;

final class ServerClientCache implements Tickable {
    use SingletonTrait;

    /** @var array<ServerClient> */
    private array $clients = [];
    /** @var array<string> */
    private array $servers = [];
    /** @var array<ServerClient> */
    private array $clientsByAddress = [];

    public function __construct() {
        self::setInstance($this);
    }

    public function add(CloudServer $server, ServerClient $client): void {
        if (!$this->isset($client)) {
            CloudLogger::get()->debug("Adding client {} => {}", $client, $server->getName());
            $this->clients[$server->getName()] = $client;
            $this->servers[$client->toString()] = $server->getName();
            $this->clientsByAddress[$client->getAddress()->toString()] = $client;
        }
    }

    public function remove(ServerClient|CloudServer $client): void {
        $client = $client instanceof CloudServer ? $this->get($client) : $client;
        if ($client !== null) {
            if ($this->isset($client)) {
                CloudLogger::get()->debug("Removing client {}", $client);
                $serverName = $this->servers[$client->toString()] ?? null;
                if ($serverName === null) {
                    unset($this->clients[array_search($client, $this->clients)]);
                } else {
                    unset($this->clients[$serverName]);
                }

                unset($this->servers[$client->toString()]);
                unset($this->clientsByAddress[$client->getAddress()->toString()]);
            }
        }
    }

    public function isset(ServerClient $client): bool {
        return isset($this->servers[$client->toString()]);
    }

    public function pick(Closure $conditionClosure): array {
        return array_filter($this->clients, $conditionClosure);
    }

    public function tick(int $currentTick): void {
        if ($currentTick % 30 == 0) {
            [$memoryUsage, $peakMemoryUsage] = array_values(ProcessUtils::getProcessStatus());
            Network::getInstance()->broadcastPacket(KeepAlivePacket::create(
                PocketCloud::getInstance()->getCurrentTPS(),
                PocketCloud::getInstance()->getAverageTPS(),
                $memoryUsage,
                $peakMemoryUsage,
                ProcessUtils::getMemoryLimit(),
                ProcessUtils::getCpuUsage()
            ));
        }

        foreach ($this->clients as $client) {
            $toRemove = [];
            foreach ($client->getDelayedPackets() as $i => $data) {
                if ($data[1] <= $currentTick) {
                    $success = $client->sendPacket($data[0]);
                    if (($data[2] ?? null) !== null) ($data[2])($client, $data[0], $success);
                    $toRemove[] = $i;
                }
            }

            foreach ($toRemove as $i) $client->unsetDelayedPacket($i);
        }
    }

    public function get(CloudServer $server): ?ServerClient {
        return $this->clients[$server->getName()] ?? null;
    }

    public function getServer(ServerClient $client): ?CloudServer {
        $serverName = $this->servers[$client->toString()] ?? null;
        if ($serverName === null) return null;
        return CloudServerManager::getInstance()->get($serverName);
    }

    public function getByAddress(Address $address): ?ServerClient {
        return $this->clientsByAddress[$address->toString()] ?? null;
    }

    public function getAll(TemplateType ...$objects): array {
        if (!empty($objects)) return array_filter($this->clients, function (ServerClient $client) use($objects): bool {
            if ($client->getServer() === null) return false;
            return array_any($objects, fn(TemplateType $object) => $object->getName() === $client->getServer()->getTemplate()->getTemplateType()->getName());
        });

        return $this->clients;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/network/client/ServerClientCache.java
package de.pocketcloud.cloud.network.client;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.network.broadcaster.PacketBroadcaster;
import de.pocketcloud.cloud.network.packet.impl.KeepAlivePacket;
import de.pocketcloud.cloud.server.CloudServer;
import de.pocketcloud.cloud.server.CloudServerManager;
import de.pocketcloud.common.lifecycle.Tickable;
import de.pocketcloud.cloud.util.PerformanceStats;
import io.netty.channel.Channel;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.net.SocketAddress;
import java.util.*;
import java.util.function.Predicate;

public final class ServerClientCache implements Tickable {

    @Getter
    @Accessors(fluent = true)
    private static ServerClientCache instance = null;

    private final Map<UUID, ServerClient> clientsByServer = new HashMap<>();
    private final Map<String, UUID> serversByClient = new HashMap<>();
    private final Map<String, ServerClient> clientsByAddress = new HashMap<>();

    public ServerClientCache() {
        instance = this;
    }

    @Override
    public void tick(long currentTick) {
        if (currentTick % 30 == 0) {
            PerformanceStats stats = PocketCloud.instance().performanceStats();
            PacketBroadcaster.broadcastPacket(KeepAlivePacket.create(
                    stats.currentTPS(),
                    stats.averageTPS(),
                    stats.processUsedMemory(),
                    stats.processPeakUsedMemory(),
                    stats.processMaxMemory(),
                    stats.processCpuUsage()
            ));
        }

        for (ServerClient client : clientsByServer.values()) {
            for (ServerClient.DelayedPacket dp : client.pollDuePackets()) {
                client.sendPacket(dp.packet())
                        .thenAccept(dp.future()::complete)
                        .exceptionally(e -> {
                            dp.future().completeExceptionally(e);
                            return null;
                        });
            }
        }
    }

    public synchronized void add(CloudServer server, ServerClient client) {
        if (isRegistered(client)) return;
        CloudLogger.get().debug("Adding client {} => {} ({})", client, server.name(), server.uuid().toString());
        clientsByServer.put(server.uuid(), client);
        serversByClient.put(client.toString(), server.uuid());
        clientsByAddress.put(client.address().toString(), client);
    }

    public synchronized void remove(ServerClient client) {
        if (!isRegistered(client)) return;
        CloudLogger.get().debug("Removing client {}", client);
        UUID serverUuid = serversByClient.remove(client.toString());
        if (serverUuid != null) {
            clientsByServer.remove(serverUuid);
        } else {
            clientsByServer.values().remove(client);
        }

        clientsByAddress.remove(client.address().toString());
    }

    public synchronized void remove(CloudServer server) {
        ServerClient client = clientsByServer.get(server.uuid());
        if (client != null) remove(client);
    }

    public synchronized boolean isRegistered(ServerClient client) {
        return serversByClient.containsKey(client.toString());
    }

    public synchronized Optional<ServerClient> getByChannel(Channel channel) {
        return Optional.ofNullable(clientsByAddress.getOrDefault(channel.remoteAddress().toString(), null));
    }

    public synchronized Optional<ServerClient> getByAddress(SocketAddress address) {
        return Optional.ofNullable(clientsByAddress.getOrDefault(address.toString(), null));
    }

    public synchronized Optional<ServerClient> get(CloudServer server) {
        return Optional.ofNullable(clientsByServer.getOrDefault(server.uuid(), null));
    }

    public synchronized CloudServer getServer(ServerClient client) {
        UUID serverUuid = serversByClient.get(client.toString());
        if (serverUuid == null) return null;
        return CloudServerManager.instance().get(serverUuid).orElse(null);
    }

    public synchronized List<ServerClient> getAll(Predicate<ServerClient> filter) {
        return clientsByServer.values().stream()
                .filter(filter)
                .toList();
    }

    public synchronized Collection<ServerClient> getAll() {
        return List.copyOf(clientsByServer.values());
    }
}